src/main/scala/uk/pkerrigan/darwin/Message.scala

```diff
@@ -0,0 +1,63 @@
+package uk.pkerrigan.darwin
+
+import scala.xml.{Node, XML}
+
+sealed trait Time
+case class EstimatedTime(time: String) extends Time
+case class ActualTime(time: String) extends Time
+
+sealed trait Location {
+  def tiploc: String
+}
+case class Station(tiploc: String, arrivalTime: Time, departureTime: Time) extends Location
+case class Waypoint(tiploc: String, passTime: Time) extends Location
+
+sealed trait Message {
+  def origin: String
+}
+case class ScheduleUpdate(origin: String, runId: String, locations: List[Location]) extends Message
+
+object Message {
+  def parseMessage(message: String): Option[Message] =
+    (XML.loadString(message) \ "uR")
+      .headOption
+      .flatMap(this.parseUpdateResponse)
+
+  private def parseUpdateResponse(updateResponse: Node): Option[Message] =
+    updateResponse.child
+      .headOption
+      .flatMap(this.parseMessageType(updateResponse.attribute("updateOrigin").head.text)(_))
+
+  private def parseMessageType(origin: String)(messageContent: Node): Option[Message] =
+    messageContent.label match {
+      case "TS" => this.parseSchedule(origin)(messageContent)
+      case _ => None
+    }
+
+  private def parseSchedule(origin: String)(messageContent: Node): Option[ScheduleUpdate] = {
+    val locations = (messageContent \ "Location")
+      .map(this.parseLocation)
+      .toList
+    Some(ScheduleUpdate(origin, messageContent.attribute("rid").head.text, locations))
```

Comment:
yep, the only difference is your function is `A => B` as opposed to `A => Option[B]`